### DISCUSSION TITLE
Change default lspconfig

### DISCUSSION THREAD

**Discussion by jeusdi** [LazyVim/LazyVim#2793] [Q&A] [answered]
According to [documentation here](https://www.lazyvim.org/extras/lang/java#nvim-jdtls) seems that I'm able to change default `root_dir` detection.

It's not clear to how to chage that. I mean, By default, it's picking default `root_dir` detection from lspconfig.

Default lspconfig [jdtls configuration is here](lspconfig.server_configurations.jdtls) and you can find [jdtls.lua lspconfig file here](https://github.com/neovim/nvim-lspconfig/blob/master/lua/lspconfig/server_configurations/jdtls.lua#L79). You can also to see how to [`lsp_config` populate it on `default_config.root_dir` here](https://github.com/neovim/nvim-lspconfig/blob/master/lua/lspconfig/server_configurations/jdtls.lua#L102).

I need to change to: 

    {'.git', 'mvnw', 'gradlew'}

Also, I've tried to change `nvim-jdtls` plugin settings using this code snipped on `_/lua/plugins/jdtls-change-root.lua`:
```lua
return {
  {
    "mfussenegger/nvim-jdtls",
    ---@type lspconfig.options.jdtls
    ---@diagnostic disable-next-line: missing-fields
    opts = {
      jdtls = function(opts)
        local root_dir = {{'.git', 'mvnw', 'gradlew'}}
        table.insert(opts.root_dir, root_dir)

        return opts
      end,
    },
  },
}
```

But something is wrong in that code.
Any ideas?

**Comment by dangdd2003** [comment, marked as answer]
As i see they are different settings, between jdtls settings from neovim and mfussengger. Neovim tries to look into file name while nvim-jdtls aims for the command. It should be configured like this (but as i changed, i didn't know the difference):
```lua
return {
  {
    "mfussenegger/nvim-jdtls",
    ---@type lspconfig.options.jdtls
    ---@diagnostic disable-next-line: missing-fields
    opts = {
      jdtls = function(opts)
        opts.root_dir = require("jdtls.setup").find_root({ ".git", "mvnw", "gradlew" })
        return opts
      end,
    },
  },
}
```

**Reply by jeusdi** [reply]
Thanks @dangdd2003 . What about changing lspconfig properties instead?

**Reply by dangdd2003** [reply]
You can check [this](https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md#jdtls) for default settings of neovim/lspconfig. `root_dir` is a funtions that loop through a string in `root_files` (you can read `return` part of file `jdtls.lua` in `server_configurations` folder). I'm not sure about neovim/lspconfig, i only config nvim-jdtls to avoid duplication. You can also see default LazyVim [config](https://www.lazyvim.org/extras/lang/java) for java. I think this will work:
```lua
{
  "neovim/nvim-lspconfig",
  opts = {
    servers = {
      jdtls = {
        -- this is the default settings in lspconfig
        root_dir = function(fname)
          local root_files = {
            -- Single-module projects
            {
              "build.xml", -- Ant
              "pom.xml", -- Maven
              "settings.gradle", -- Gradle
              "settings.gradle.kts", -- Gradle
            },
            -- Multi-module projects
            { "build.gradle", "build.gradle.kts" },
          }
          for _, patterns in ipairs(root_files) do
            local root = require("lspconfig.util").root_pattern(unpack(patterns))(fname)
            if root then
              return root
            end
          end
        end,
      },
    },
  },

}
```
